feat: Adds support for Cursor provider and other ACP backends

aweis89 opened the pull request:
## Summary
- While ACP is already supported for custom IDE integrations (`IDE/ACP-> OpenCode -> LLM`), this adds support to use ACP as a backend LLM provider (`OpenCode -> ACP`)
-  Enables OpenCode to consume tokens from Cursor CLI for example while still using the regular OpenCode TUI.
- This is useful for folks who have Cursor subscriptions for example, but want to use OpenCode's (amazing!) TUI
- By enabling ACP as a backend provider, we don't need a uniq translation layer for OpenCode to multiple CLI backends
 
## Resolves
- Closes #3735 - CLI Providers feature request
- Closes #2072 - Support for Cursor CLI
- Closes #402 - Gemini CLI support for free tier access
- Closes #5182


## Implementation Details

### Core Components
1. **ACPClient** (`client.ts`) - Manages subprocess lifecycle and ACP protocol communication
2. **ACPLanguageModel** (`model.ts`) - Implements LanguageModelV2 interface with streaming support
3. **Message Converters** (`converters.ts`) - Bidirectional conversion between Vercel AI SDK and ACP message formats
4. **MCP Converter** (`mcp-converter.ts`) - Converts OpenCode MCP configurations to ACP format
5. **Provider Factory** (`factory.ts`) - Creates ACP provider instances from configuration

### Key Features
- Full streaming support for real-time responses
- Tool execution, passes along MCP servers, which are executed by the ACP
- MCP server integration - OpenCode MCP servers exposed to ACP backends
- Proper subprocess lifecycle management and cleanup
- Comprehensive error handling

## Testing
- Tested with cursor-agent integration
- Verified streaming responses
- Confirmed MCP server exposure to ACP backend

## Documentation
- Added comprehensive README.md with setup instructions
- Includes configuration examples for cursor-agent, goose, and gemini-cli
- Documents architecture and design decisions

rekram1-node commented:
we already have acp support, whats this for

aweis89 commented:
> we already have acp support, whats this for

This is for an ACP backend provider. For example, if you have a cursor account, and you want to use Opencode TUI's interface, but you want it to use cursor CLI agent as it's backend LLM:

<img width="3024" height="1964" alt="Screenshot 2025-12-04 at 7 37 02 PM" src="https://github.com/user-attachments/assets/5ff2566b-8458-4c34-ac23-a68c5e19a6a9" />

Right now we have: `IDE via ACP -> OpenCode -> LLM Provider`
With this change you can now do: `OpenCode -> Cursor CLI via ACP`
The ACP becomes the LLM provider as apposed to the frontend interface.
This is useful for folks who want to use OpenCode but already have a subscription for cursor.
They can now use OpenCode while comsuming their tokens from Cursor

rekram1-node commented:
okay i see

aweis89 commented:
> okay i see

Been using this all morning. It's pretty awesome to be able to use my Cursor subscription with OpenCode's best-in-class TUI!

aweis89 committed:
feat: adds support for backed ACP provider

aweis89 committed:
Merge branch 'dev' into enable-acp

aweis89 committed:
fix: allow unknown tool calls for ACP providers

Implemented a Proxy for the tools record when an ACP provider is active.
This dynamically provides pass-through tool definitions for unknown
tool calls, allowing them to be correctly proxied to the ACP.

aweis89 committed:
Merge branch 'dev' into enable-acp

aweis89 committed:
Merge branch 'dev' into enable-acp

aweis89 committed:
Merge branch 'dev' into enable-acp

aweis89 commented:
<img width="1503" height="909" alt="Screenshot 2025-12-20 at 8 29 02 PM" src="https://github.com/user-attachments/assets/11a45448-0d8b-4c04-aee1-c6156b600f6d" />

aweis89 commented:
Any chance this is something we can merge @rekram1-node?

aweis89 committed:
Merge branch 'dev' into enable-acp

jensenojs commented:
same ask here

aweis89 committed:
chore: merge dev branch to resolve conflicts

Resolved conflicts in:
- transform.ts: Added @ai-sdk/google-vertex fall-through case while keeping ACP context logic
- llm.ts: Merged ACP tool/jsonSchema imports with extractReasoningMiddleware from dev

aweis89 committed:
Merge branch 'dev' into enable-acp

aweis89 committed:
Merge branch 'dev' into enable-acp

aweis89 committed:
Merge branch 'dev' into enable-acp

Nasseratic commented:
Awesome work @aweis89  ❤️

NazmiAltun commented:
@rekram1-node  any plans to have this feature ?

aweis89 committed:
Merge branch 'dev' into enable-acp

dleggo commented:
I noticed when trying this out with gemini-cli that gemini does not respect the permission settings in my `opencode.jsonc`.

```json
{
  "provider": {
    "gemini-acp-provider": {
      "name": "Gemini Cli",
      "models": {
        "auto": {
          "id": "gemini-3-pro-preview",
          "name": "Gemini 3 Pro"
        }
      },
      "options": {
        "command": "gemini",
        "args": [
          "--experimental-acp"
        ]
      }
    }
  },
  "$schema": "https://opencode.ai/config.json",
  "agent": {
    "build": {
      "permission": {
        "edit": "ask"
      }
    }
  },
  "autoupdate": false
}
```

Is this a limitation of the LLM being provided via ACP @aweis89 ?
Regular models do respect this setting.

aweis89 commented:
> I noticed when trying this out with gemini-cli that gemini does not respect the permission settings in my `opencode.jsonc`.
> 
> ```json
> {
>   "provider": {
>     "gemini-acp-provider": {
>       "name": "Gemini Cli",
>       "models": {
>         "auto": {
>           "id": "gemini-3-pro-preview",
>           "name": "Gemini 3 Pro"
>         }
>       },
>       "options": {
>         "command": "gemini",
>         "args": [
>           "--experimental-acp"
>         ]
>       }
>     }
>   },
>   "$schema": "https://opencode.ai/config.json",
>   "agent": {
>     "build": {
>       "permission": {
>         "edit": "ask"
>       }
>     }
>   },
>   "autoupdate": false
> }
> ```
> 
> Is this a limitation of the LLM being provided via ACP @aweis89 ? Regular models do respect this setting.

Yes, ACP is designed to have the backend ACP provider execute the tools on it's own. This is unlike a regular model which delegates tool execution to OpenCode. Therefore OpenCode cannot enforce which tools get executed and this needs to be configured via the backend provider instead. I should probably document this as a key limitation when using an ACP backend.

wzj-zz commented:
cool！！！

hawkff commented:
This is actually awesome!

crper commented:
awesome!!!

kaan-zeroday commented:
awesome

smnatale commented:
can confirm, this is pretty cool 
<img width="852" height="1050" alt="image" src="https://github.com/user-attachments/assets/1ef57f1e-f9a7-41a5-b366-ed07787ed90b" />

helderdiin commented:
Great job! I'm excited to use it haha

fwang committed:
wip: zen

fwang committed:
wip: zen

fwang committed:
wip: black

actions-user committed:
chore: generate

fwang committed:
wip: black

dbpolito committed:
feat(desktop): Change Terminal's Cursor to Bar and Stop Blinking when not focused (#7402)

fathiraz commented:
this is very cool 🚀 !

fwang committed:
wip: black

fwang committed:
wip: black

actions-user committed:
chore: generate

rekram1-node committed:
fix: model dialog issue when searching for models in fav/recents list, also ensure that deprecated models dont appear in list (#7429)

aweis89 committed:
Merge branch 'dev' into enable-acp

aweis89 committed:
Merge upstream/dev into enable-acp

Resolved conflict in packages/console/app/src/routes/stripe/webhook.ts:
- Added isNull import from upstream/dev
- Kept removal of timeSubscribed update from HEAD
- Added new invoice.payment_succeeded and charge.refunded handlers from upstream/dev

aweis89 committed:
fix: update ACP client for new permission system

- Use PermissionNext.evaluate() instead of accessing permission config
  properties directly (bash, edit, external_directory)
- Change Permission.ask() parameter from 'title' to 'message'
- Update handlePermissionAction to use PermissionNext.Action type
- Fix MCP converter to handle { enabled: boolean } config entries

aweis89 committed:
docs(acp): clarify that permissions cannot be enforced with ACP backends

When using ACP as a backend provider, tool execution is handled by the
ACP subprocess which has direct filesystem/network access. OpenCode can
only request permission via callbacks - actual enforcement depends on
whether the ACP backend chooses to respect the response.

aweis89 committed:
fix: remove batch onboard script

aweis89 committed:
prevent detached on Windows to avoid issues with process management

aweis89 committed:
remove hardcoded version number in acp client.ts

aweis89 committed:
better error handling for permission denials

aweis89 committed:
remove entire model list for ACP provider in opencode.jsonc

aweis89 committed:
use promise instead of polling to wait for pending notifications

aweis89 committed:
Update README to remove future enhancements

Removed future enhancements section from README.

aweis89 committed:
add types for ACP tool results and improve logging

aweis89 committed:
revert unecessary await

aweis89 committed:
document addinng type: acp to ACP provider README.md

aweis89 committed:
Merge branch 'dev' into enable-acp

oraclian94 commented:
I hope this PR gets merged soon!

aweis89 committed:
fix(acp): normalize tool input field names for edit tool
- update normalization to handle 'path' for both read and edit tools

aweis89 committed:
Merge branch 'dev' into enable-acp

aweis89 committed:
Retry wrapped ECONNRESET errors

Treat wrapped "ConnectError: ... ECONNRESET" (common via ACP subprocess backends) as retryable by unwrapping cause/message, and add a regression test.

LeonMueller-OneAndOnly commented:
Here is another interesting approach for getting cursor to work in opencode

https://github.com/R44VC0RP/cursor-opencode-auth

aweis89 committed:
Merge remote-tracking branch 'upstream/dev' into enable-acp

aweis89 committed:
Merge remote-tracking branch 'upstream/dev' into enable-acp

aweis89 committed:
Merge upstream/dev into enable-acp

Nomadcxx commented:
> This isn't something we have bandwidth to maintain at the moment, this may be better suited to be written as a plugin 

Noooooooooo

kaan-zeroday commented:
@aweis89 Can you publish as plugin?
